archive/100-theorems-list/57_herons_formula.lean

```diff
@@ -0,0 +1,108 @@
+/-
+Copyright (c) 2020 Matt Kempster. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Matt Kempster
+-/
+import geometry.euclidean.triangle
+
+/-!
+# Freek № 57: Heron's Formula
+
+This file proves Theorem 57 from the [100 Theorems List](https://www.cs.ru.nl/~freek/100/),
+also known as Heron's formula, which gives the area of a triangle based only on its three sides'
+lengths.
+
+## References
+
+* https://en.wikipedia.org/wiki/Herons_formula
+
+-/
+
+open euclidean_geometry
+open_locale real euclidean_geometry
+
+notation `√` := real.sqrt
+
+variables {V : Type*} {P : Type*} [inner_product_space ℝ V] [metric_space P]
+    [normed_add_torsor V P]
+
+/-- The sine of a real number `x` is the square root of one minus the cosine squared of `x`. -/
+lemma sin_eq_sqrt_one_minus_cos_sq (x : ℝ) (hl : x ≥ 0) (hu : x ≤ π) :
+  real.sin x = √ (1 - (real.cos x)^2) :=
+begin
+  rw ← real.sqrt_sqr (real.sin_nonneg_of_nonneg_of_le_pi hl hu),
+  exact congr_arg real.sqrt (real.sin_square x),
+end
+
+lemma rearrange_cos_rule (a b c d : ℝ) (a_nonzero : a ≠ 0) (b_nonzero : b ≠ 0):
+  c * c = a * a + b * b - 2 * a * b * d → d = (a * a + b * b - c * c) / (2 * a * b) :=
+begin
+  intro hyp,
+  simp only [*, sub_sub_cancel],
+  field_simp,
+  ring,
+end
```

Comment:
By moving your last assumption before the colon, this becomes much simpler:
```suggestion
lemma rearrange_cos_rule (a b c d : ℝ) (a_nonzero : a ≠ 0) (b_nonzero : b ≠ 0)
  (h : c * c = a * a + b * b - 2 * a * b * d) : d = (a * a + b * b - c * c) / (2 * a * b) :=
by field_simp [h, mul_comm]
```
As you can see, it is now simple enough that it probably should not be its own lemma.